### PR TITLE
Fix for #175 Remove unnecessary msg.Clone calls

### PR DESCRIPTION
**Pull request by arnecls**
Referring to #175

This will remove msg.Clone from producers perviously cloning the original message for retry.
As RetryMessage already relies on the original stored within a message this is not necessary anymore.
In addition to this, all producers that formerly used specific key formatters have ben adjusted to retrieve the key from metadata.

### CODE REVIEW

File: core/message.go [hunk 1]
```diff
@@ -211,6 +212,21 @@ func (msg Message) Serialize() ([]byte, error) {
...
+		StreamID:     proto.Uint64(uint64(msg.orig.streamID)),
+		PrevStreamID: proto.Uint64(uint64(msg.data.streamID)),
+		Timestamp:    proto.Int64(msg.timestamp.UnixNano()),
+		Data:         msg.orig.payload,
+		Metadata:     msg.orig.Metadata,
+	}
+
+	return proto.Marshal(serializable)
+}
+
```
Comment (msiebeneicher) [inline core/message.go]: please provide a `func serializeMsg(msg *Messsage) ([]byte, error)` method to remove this duplication

File: producer/redis.go [hunk 2]
```diff
@@ -54,34 +51,23 @@ import (
...
+// this value is set to an empty string.
 type Redis struct {
 	core.BufferedProducer `gollumdoc:"embed_type"`
 	address               string
 	protocol              string
 	password              string `config:"Password"`
 	database              int    `config:"Database" default:"0"`
-	key                   string `config:"Key" default:"default"`
+	keyField              string `config:"KeyFrom"`
+	fieldField            string `config:"FieldFrom"`
```
Comment (msiebeneicher) [inline producer/redis.go]: bad wording - `fieldField`sounds like a typo issue

File: producer/s3.go [hunk 3]
```diff
@@ -520,8 +520,6 @@ func (prod *S3) transformMessages(messages []*core.Message) {
 	bufferedMessages := []*core.Message{}
 	// Format and sort
 	for _, msg := range messages {
-		originalMsg := msg.Clone()
```
Comment (msiebeneicher) [inline producer/s3.go]: merge conflict because of #173 - reverting this file should resolve the merge conflict

File: producer/redis.go (hunk 2 above)
Comment (arnecls) [inline producer/redis.go]: :D yeah.

File: core/message.go (hunk 1 above)
Comment (arnecls) [inline core/message.go]: 👍

File: producer/s3.go (hunk 3 above)
Comment (arnecls) [inline producer/s3.go]: Interestingly there is no conflict when I merge this ...

**Pull request merged**